libs/futures/include/hpx/lcos/local/futures_factory.hpp

```diff
@@ -124,24 +118,39 @@ namespace hpx { namespace lcos { namespace local {
                         threads::pending_do_not_schedule, true,
                         threads::thread_priority_boost,
                         threads::thread_schedule_hint(
+                            // TODO: Replace with the internal access functions.
                             static_cast<std::int16_t>(get_worker_thread_num())),
                         stacksize, ec);
                 }
 
                 threads::register_thread_nullary(pool,
                     util::deferred_call(&base_type::run_impl, std::move(this_)),
-                    util::thread_description(f_, annotation),
-                    threads::pending, false, priority, schedulehint, stacksize,
-                    ec);
+                    util::thread_description(f_, annotation), threads::pending,
+                    false, priority, schedulehint, stacksize, ec);
+                return threads::invalid_thread_id;
+            }
+
+            template <typename Spawner>
+            threads::thread_id_type apply(
+                Spawner&& spawner, error_code& ec)
+            {
+                this->check_started();
+
+                typedef typename Base::future_base_type future_base_type;
+                future_base_type this_(this);
+
+                spawner(
+                    util::deferred_call(&base_type::run_impl, std::move(this_)));
+
```

Comment:
@hkaiser I've made `futures_factory` take a "spawner" instead of relying directly on any particular method of spawning tasks (`register_thread_nullary`, `post` etc.). This works nicely to make this part of futures independent of execution (the `post` customization point) and `register_thread` (in principle not a problem, but does not need to be a dependency). It also removes the special case of handling thread executors differently.

`packaged_continuation` has similar dependencies on `post` and `post_policy_dispatch` for `async_exec` and `async`. Do you see a problem with giving `packaged_continuation` the same treatment as `task_object`? `packaged_continuation` could then live with the rest of the futures files and the `make_continuation*` wrappers would then live in the execution module.

As an alternative the `post` etc. customization points and `post_policy_dispatch`/`async_launch_policy_dispatch` declarations could move to the `async_base` or `basic_execution` module. This would mean that all the different ways of spawning tasks are available for the futures module. In that case I can also revert this particular change to `futures_factory`.

Do you have a preference or other suggestion?

Reply:
@msimberg I like the change to `futures_factory`, that makes it much more reusable. I have no objections to adapt `packaged_continuation` either. Also I like the move of the launching facilities to `async_base`. Thanks!

Reply:
I should test my changes properly before saying they work nicely. The change to make `futures_factory::apply` take a templated spawner does not work because it would have to go through a templated member function in `task_base` (which is not allowed). Because of performance concerns I would not like to use `hpx::function` (or `unique_function_nonser`), but this concern may be unfounded?

I've reverted that change now as we can go ahead without that change. I've made `packaged_continuation` take a generic spawner though, and I've moved the execution customization points to `basic_execution`.

Reply:
> The change to make futures_factory::apply take a templated spawner does not work because it would have to go through a templated member function in task_base (which is not allowed).

Why wouldn't that be allowed?

Reply:
> > The change to make futures_factory::apply take a templated spawner does not work because it would have to go through a templated member function in task_base (which is not allowed).
> 
> Why wouldn't that be allowed?

I left out a word: *virtual* template member function. There may be another way around it though...

Reply:
Since this doesn't cause circular dependencies and removing thread executors will get rid of that dependency anyway I'm going to leave things as they are in `futures_factory`.